In [17]:
#reading multiple csv files
import pandas as pd
import glob #glob is amazing

csv_files = glob.glob(r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\*.csv") #reads any file in the directory that ends with .csv

datasets = {}

for file in csv_files: 
    data = pd.read_csv(file) #read each file one by one
    datasets[file] = data #put each file into the dictionary so we can call those files without having to read each one again



In [18]:
#inspecting customers dataset

customers = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_customers_dataset.csv"]

customers.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


**Customer id, customer_unique_id** look like an idetifier. not human readable.

Questions: 
- "is it unique?"
- "does it appear in other tables?"
- "does it have a consistent format?"
- "are there missing values?"
- "are there duplicates?"
- "why would a table need 2 ids?:

**Customer_zip_code_prefix** - in the data head itself i notice two values 1151 and 4534 having fewer than 5 digits. 

lets inpect another table 

In [19]:
#inspectin orders_dataset

orders = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_orders_dataset.csv"]

orders.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


**order_status** is important, it describes the state of an order
the date columns need to be examined for consistency and datatype
NaN values need to be quantified, not just noticed visually
the different date columns probably have difference business meaning so we shouldnt assume they all behave identically

In [20]:
orders.info()
orders.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

**IN ORDERS DATASET**      

- `customer_id` appears in both the customers and orders tables.
- `order_id` is present in the orders table.
- Several timestamp/date columns are present.
- Some date columns contain missing values.
- All columns are currently stored as strings.


**MISSING VALUES:**

`order_approved_at` = 99441 - 99281 = 160

`order_delivered_carrirer_date` 99441 - 97658 =  1783

`order_delivered_customer_date` 99441 - 96476 = 2965


**Questions:**

- "How do we fill these NaNs?"
- "Why are these values missing?"
- "Are missing delivery dates associated with "order status" values?"
- "Are all timestamp columns consistently formatted?"

if an order hasnt been delivered, it might naturally have no value.

In [21]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Missing delivery date investigation

- `order_status` contains 96,478 `delivered` orders out of 99,441 total orders.
- `order_delivered_customer_date` has 2,965 missing values.
- There are 2,963 non-delivered orders:
  `99,441 - 96,478 = 2,963`
- Therefore, 2 missing `order_delivered_customer_date` values appear to belong to orders marked as `delivered`.

### Question

- Why do these 2 orders have `order_status = delivered` but no `order_delivered_customer_date`?
- Need to inspect the 2 records before deciding whether they are data-quality errors.

In [22]:
orders[orders["order_delivered_customer_date"].isnull() & (orders["order_status"] == "delivered")]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


In [23]:
pd.crosstab(
    orders["order_status"],
    orders["order_delivered_customer_date"].isnull()
)

order_delivered_customer_date,False,True
order_status,,
approved,0,2
canceled,6,619
created,0,5
delivered,96470,8
invoiced,0,314
processing,0,301
shipped,0,1107
unavailable,0,609


### Missing delivery date investigation

- `order_delivered_customer_date` has 2,965 missing values.
- Filtering for orders where:
  - `order_status == "delivered"`
  - `order_delivered_customer_date` is missing
- returned 8 records.
- Therefore, missing delivery dates are not exclusively associated with non-delivered orders.

### Question

- Why are these 8 orders marked as `delivered` despite having no customer delivery date?
- Need to investigate these records and determine whether they represent legitimate missing data or data-quality issues.

In [24]:
orders["order_purchase_timestamp"].tail(20)


99421    2017-07-05 15:55:19
99422    2018-05-25 17:05:17
99423    2017-10-12 20:54:11
99424    2017-11-21 20:18:03
99425    2018-03-21 19:47:18
99426    2017-05-15 09:46:26
99427    2018-05-12 13:42:24
99428    2018-06-22 20:53:29
99429    2017-11-05 16:17:53
99430    2018-03-04 22:48:38
99431    2017-10-27 15:21:00
99432    2017-12-20 09:52:41
99433    2017-10-04 19:57:37
99434    2017-01-27 00:30:03
99435    2017-02-23 09:05:12
99436    2017-03-09 09:54:05
99437    2018-02-06 12:58:58
99438    2017-08-27 14:46:43
99439    2018-01-08 21:28:27
99440    2018-03-08 20:57:30
Name: order_purchase_timestamp, dtype: str

In [30]:
pd.to_datetime(orders["order_purchase_timestamp"], errors = 'coerce')
orders["order_purchase_timestamp"].isnull().sum()

np.int64(0)

### `order_purchase_timestamp`

- Stored as `str` in the raw dataset.
- Values follow the apparent format `YYYY-MM-DD HH:MM:SS`.
- `pd.to_datetime(..., errors="coerce")` successfully parsed all 99,441 values.
- 0 values became `NaT`.
- No invalid/unparseable values detected in this column.

In [36]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [37]:
date_columns = ['order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date']

for column in date_columns:
    converted = pd.to_datetime(orders[column], errors = "coerce")
    print(column, converted.isnull().sum())



order_purchase_timestamp 0
order_approved_at 160
order_delivered_carrier_date 1783
order_delivered_customer_date 2965
order_estimated_delivery_date 0


### Date columns

All date columns are stored as `str` in the raw dataset.

After testing conversion with `pd.to_datetime(..., errors="coerce")`:

| Column | Failed/NaT values |
|---|---|
| `order_purchase_timestamp` | 0 |
| `order_approved_at` | 160 |
| `order_delivered_carrier_date` | 1,783 |
| `order_delivered_customer_date` | 2,965 |
| `order_estimated_delivery_date` | 0 |

Observations:
- No invalid date strings were detected.
- The missing values in the three affected columns remain missing after conversion.
- Date values appear to follow a consistent `YYYY-MM-DD HH:MM:SS` format.
- These columns can likely be converted to datetime during the cleaning phase.